Fix type error in token comparison with python 2.7

mrvn opened the pull request:
I'm using this simple code to test python-oath:

import oath
from Crypto import Random
import binascii
day = 24 * 60 * 60
key = binascii.hexlify(Random.get_random_bytes(16)).decode('ascii')
print("key = {0}".format(key))
token = oath.totp(key, format='hex-notrunc', period = day)
print("token = {0} {1}".format(token, type(token)))
(ok, drift) = oath.accept_totp(key, response, format='hex-notrunc', period = day, backward_drift = 2)
print("ok = {0}, drift = {1}".format(ok, drift))

but this results in a TypeError under python 2.7:

python test.py 
key = c2b0bd28f9d9da8c7375fa6321448397
token = b767fe972a7e4626b2c7ad578c4b1fa4ccd29f1f \<type 'unicode'\>
Traceback (most recent call last):
  File "test.py", line 15, in <module>
    (ok, drift) = oath.accept_totp(key, response, format='hex-notrunc', period = day, backward_drift = 2)
  File "/usr/lib/python2.7/dist-packages/oath/_totp.py", line 129, in accept_totp
    if _utils.compare_digest(totp(key, format=format, period=period, hash=hash, t=t+d), str(response)):
  File "/usr/lib/python2.7/dist-packages/oath/_utils.py", line 27, in compare_digest
    raise TypeError
TypeError

bdauvergne commented:
That's not the correct fix for the problem. totp should never return an unicode string. What happened is that for python 3.2 support, a .decode('ascii') call was appended to an binascii.hexlify() call since with python 3.2 it returns bytes instead of strings. I think we need our own hexlify() function which would always return an str, i.e. under python 2 and 3.

mrvn commented:
From the comments I gathered that .decode('ascii') was supposed to return the same type in python 2.7 and 3.2. But that just changes str/byte to unicode/str so the type problem there remains. But I think that is a different problem that just happens to trigger the bug above and should get it's own issue because it breaks compatibility when different python versions are used to create and verify tokens.

The problem here is that in the compare_digest() call one fo the arguments is converted using str() and the other isn't. The str() call used to be a NOP but due to the type problem with hexlify it isn't currently. Adding str() to the first maybe isn't the right fix. Why is it used for the second argument at all? If that is dropped then the compare_digest would just compare two 'unicode' and that would work too.

mrvn commented:
The _utils module already had a python version aware fromhex function. I've added a tohex function there too and changed all uses of binasci.hexlify to _utils.tohex and binascii.unhexlify to _utils.fromhex.

bdauvergne commented:
Great, thanks.
